crypto/x509v3/v3_ncons.c

```diff
@@ -297,47 +297,152 @@ int NAME_CONSTRAINTS_check(X509 *x, NAME_CONSTRAINTS *nc)
 
 }
 
+#define CN2DNSID_NOMEM   (-2)
+#define CN2DNSID_BADNAME (-1)
+
+static int cn2dnsid(ASN1_STRING *cn, unsigned char **dnsid)
+{
+    int utf8_length;    /* Return type of ASN1_STRING_to_UTF8 */
+    int i;
+    unsigned char *utf8_value;
+    int isdnsname = 0;
+
+    /*
+     * Per RFC 6125 DNS-IDs representing internationalized domain names appear
+     * in certificates in A-label encoded form.  The same applies to CNs which
+     * are intended to represent DNS names.  However, while in the SAN DNS-IDs
+     * are IA5Strings as CNs they may be needlessly stored in 16-bit Unicode.
+     * We perform a conversion to UTF-8 here, to ensure that we get an ASCII
+     * representation of strings that have been encoded in some other way in
+     * the CN.
+     *
+     * Note, 'int' is the return type of ASN1_STRING_to_UTF8() so that's what
+     * we must use for 'utf8_length'.
+     */
+    if ((utf8_length = ASN1_STRING_to_UTF8(&utf8_value, cn)) < 0)
+        return CN2DNSID_NOMEM;
+
+    /*
+     * Some certificates have had names that include a *trailing* NUL byte.
+     * Remove these harmless NUL characters. They would otherwise yield false
+     * alarms with the following embedded NUL check.
+     */
+    while (utf8_length > 0 && utf8_value[utf8_length-1] == '\0')
+        --utf8_length;
+
+    /* Reject *embedded* NULs */
+    if ((size_t)utf8_length != strlen((char *)utf8_value))
+        return CN2DNSID_BADNAME;
+
+    /*
+     * XXX: Deviation from strict DNS name syntax, also check names with '_'
+     * Check DNS name syntax, any '-' or '.' must be internal,
+     * and on either side of each '.' we can't have a '-' or '.'.
+     *
+     * If the name has just one label, we don't consider it a DNS name.  This
+     * means that "CN=sometld" cannot be precluded by DNS name constraints, but
+     * that is not a problem.
+     */
+    for (i = 0; i < utf8_length; ++i) {
+        unsigned char c = utf8_value[i];
+
+        if ((c >= 'a' && c <= 'z')
+            || (c >= 'A' && c <= 'Z')
+            || (c >= '0' && c <= '9')
+            || c == '_')
+            continue;
```

Comment:
There is kind of ambiguity between CN conversion to UTF8 and this check. I mean if non-ASCII characters are permitted to appear in CN, and host does have non-ASCII in its name, then would it be reasonable to expect that it's "human-friendly" one? Of course RFC says that subjectAltName is the recommended one, CN-based is kind of for backward compatibility. And one can as well argue that non-ASCII chars in hostname are not *that* backward compatible...